File: app/processing/crop_transform.py
from typing import Tuple
import numpy as np
import cv2

def crop_image_and_mask(
    image: np.ndarray,
    mask: np.ndarray,
    crop_rect: Tuple[int, int, int, int]
) -> Tuple[np.ndarray, np.ndarray]:
    """Crops image and alpha mask using bounding box tuple (x, y, width, height)."""
    if image is None or crop_rect is None:
        return image, mask

    x, y, w, h = crop_rect
    img_h, img_w = image.shape[:2]

    x1 = max(0, min(x, img_w - 1))
    y1 = max(0, min(y, img_h - 1))
    x2 = max(x1 + 1, min(x1 + w, img_w))
    y2 = max(y1 + 1, min(y1 + h, img_h))

    cropped_img = image[y1:y2, x1:x2].copy()
    cropped_mask = mask[y1:y2, x1:x2].copy() if mask is not None else None

    return cropped_img, cropped_mask

def apply_transforms(
    image: np.ndarray,
    scale: float = 1.0,
    rotation: float = 0.0,
    flip_h: bool = False,
    flip_v: bool = False
) -> np.ndarray:
    """Applies rotation, scaling, and horizontal/vertical flips to image array."""
    if image is None or image.size == 0:
        return image

    result = image.copy()

    if flip_h:
        result = cv2.flip(result, 1)
    if flip_v:
        result = cv2.flip(result, 0)

    if rotation != 0.0:
        h, w = result.shape[:2]
        center = (w // 2, h // 2)
        matrix = cv2.getRotationMatrix2D(center, rotation, scale)
        result = cv2.warpAffine(result, matrix, (w, h), flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_CONSTANT, borderValue=0)
    elif scale != 1.0:
        h, w = result.shape[:2]
        new_w = max(1, int(w * scale))
        new_h = max(1, int(h * scale))
        result = cv2.resize(result, (new_w, new_h), interpolation=cv2.INTER_LINEAR)

    return result

File: src/processing/CropTransform.cpp
#include "processing/CropTransform.h"
#include <algorithm>

namespace ImageCut {
namespace Processing {

std::tuple<cv::Mat, cv::Mat> CropTransform::cropImageAndMask(
    const cv::Mat& image,
    const cv::Mat& mask,
    const cv::Rect& cropRect
) {
    if (image.empty() || cropRect.width <= 0 || cropRect.height <= 0) {
        return std::make_tuple(image, mask);
    }

    int imgW = image.cols;
    int imgH = image.rows;

    int x1 = std::max(0, std::min(cropRect.x, imgW - 1));
    int y1 = std::max(0, std::min(cropRect.y, imgH - 1));
    int x2 = std::max(x1 + 1, std::min(x1 + cropRect.width, imgW));
    int y2 = std::max(y1 + 1, std::min(y1 + cropRect.height, imgH));

    cv::Rect validRoi(x1, y1, x2 - x1, y2 - y1);
    cv::Mat croppedImg = image(validRoi).clone();
    cv::Mat croppedMask = !mask.empty() ? mask(validRoi).clone() : cv::Mat();

    return std::make_tuple(croppedImg, croppedMask);
}

cv::Mat CropTransform::applyTransforms(
    const cv::Mat& image,
    double scale,
    double rotation,
    bool flipH,
    bool flipV
) {
    if (image.empty()) return image;

    cv::Mat result = image.clone();

    if (flipH && flipV) {
        cv::flip(result, result, -1);
    } else if (flipH) {
        cv::flip(result, result, 1);
    } else if (flipV) {
        cv::flip(result, result, 0);
    }

    if (rotation != 0.0) {
        int h = result.rows;
        int w = result.cols;
        cv::Point2f center(w / 2.0f, h / 2.0f);
        cv::Mat rotMatrix = cv::getRotationMatrix2D(center, rotation, scale);
        cv::warpAffine(result, result, rotMatrix, cv::Size(w, h), cv::INTER_LINEAR, cv::BORDER_CONSTANT, cv::Scalar(0, 0, 0, 0));
    } else if (scale != 1.0) {
        int newW = std::max(1, static_cast<int>(result.cols * scale));
        int newH = std::max(1, static_cast<int>(result.rows * scale));
        cv::resize(result, result, cv::Size(newW, newH), 0, 0, cv::INTER_LINEAR);
    }

    return result;
}

} // namespace Processing
} // namespace ImageCut
